java/client/test/org/openqa/selenium/ShadowElementFinderTest.java

```diff
@@ -0,0 +1,113 @@
+package org.openqa.selenium;
+
+import org.junit.Test;
+import org.openqa.selenium.testing.Ignore;
+import org.openqa.selenium.testing.JUnit4TestBase;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.openqa.selenium.testing.drivers.Browser.FIREFOX;
+import static org.openqa.selenium.testing.drivers.Browser.MARIONETTE;
+
+/**
+ * Shadow Root is not supported by Firefox.
```

Comment:
Shadow-DOM is supported in Firefox https://developer.mozilla.org/en-US/docs/Web/API/ShadowRoot#Browser_compatibility

But calling "element.shadowRoot" will return the map of different attributes, so you cannot cast it to WebElement but must directly search inside it by "shadowRoot.querySelector()"

Reply:
Makes sense, and it's possible.  
But to get the selector out of `By`, I would have to use a `toString()` and remove the prefix `By.cssSelector: ` manually, which may cause issues later if `toString()` is changed to something else.  
A better option would be to add a method to `By` class to get only the selector.  
I will make some changes

Reply:
@dstekanov Please take a look on the new implementation.